src/models/base_model.py

```diff
@@ -0,0 +1,245 @@
+
+    Example:
+        >>> class ModelWithHooks(BaseModel):
+        >>>     def __init__(self, input_channel: int):
+        >>>         super().__init__()
+        >>>         out_features = [5, 10, 15]
+        >>>         self.conv1 = nn.Conv2d(input_channel, 5, 3)
+        >>>         self.conv2 = nn.Conv2d(5, 10, 3)
+        >>>         self.conv3 = nn.Conv2d(10, 15, 3)
+        >>>         # Call _create_hooks after all modules is initialize, with parameters as in _get_features_info.
+        >>>         self._create_hooks(output_channels=[10, 15], module_names=['conv2', 'conv3'])
+
+        >>>     def _get_features_info(self, output_channels, module_names):
+        >>>         # Overwrite get_features_info.
+        >>>         features_info = []
+        >>>         for num_channels, module_name in zip(output_channels, module_names):
+        >>>             feature_info = FeatureInfo(module_name=module_name, num_channels=num_channels, stride=1)
+        >>>             features_info.append(feature_info)
+        >>>         return features_info
+    """
+    def __init__(self):
+        """Initialize BaseModel.
+        
+        Set features_info and feature_hooks as None.
+        """
+        super().__init__()
+        self._features_info = None
+        self._feature_hooks = None
+
+    def _get_features_info(self) -> List[FeatureInfo]:
+        """Method for initialize Hooks.
+        
+        It should be overriden in an inherited class if a user is expected to have access to feature hooks 
+        of the implemented Module.
+
+        Retruns: Hooks FeatureInfo list.
+        """
+        return None
+
+    def _create_hooks(self) -> Tuple[List[FeatureInfo], FeatureHooks]:
+        """Call self._get_features_info() method to generate features_info list and then generate feature hooks.
+        
+        If self._get_features_info() isn't overrided in an inherited class then the features_info and hooks 
+        will be None and no feature hooks will be created.
+
+        Args:
+            args: Any list which will be passed to self._get_features_info function.
+            kwargs: Any dict which will be passed to self._get_features_info function.
+
+        Returns:
+            features_info: Hooks feature info list.
+            feature_hooks: Generated hooks.
+
+        Raises:
+            ValueError: If overriden self._get_features_info() method return not FeatureInfo list.
+        """
+        self._features_info = self._get_features_info()
+        # Check if all self._features_info is FeatureInfo types.
+        if self._features_info is not None:
+            self.__check_features_info_types()
+            hooks = [
+                Hook(module_name=feature.module_name, hook_type=feature.hook_type) 
+                for feature in self._features_info
+                ]
+            self._feature_hooks = FeatureHooks(hooks, self.named_modules())
+
+    def __check_features_info_types(self):
```

Comment:
Make it static and pass `features_info` as argument. Reason: if things might go bad with types then `self._features_info` shouldn't be set beforehand